drivers/mdio/mdio_stm32_hal.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024 BayLibre, SAS
+ * Copyright (c) 2010-2024 Analog Devices Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/mdio.h>
+#include <zephyr/net/ethernet.h>
+#include <zephyr/net/mdio.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mdio_stm32_hal, CONFIG_MDIO_LOG_LEVEL);
+
+#define DT_DRV_COMPAT st_stm32_mdio
+
+struct mdio_stm32_data {
+	struct k_sem sem;
+	ETH_HandleTypeDef hmdio;
+};
+
+struct mdio_stm32_config {
+	const struct pinctrl_dev_config *pincfg;
+};
+
+static int mdio_stm32_read(const struct device *dev, uint8_t prtad,
+			   uint8_t regad, uint16_t *data)
+{
+	struct mdio_stm32_data *const dev_data = dev->data;
+	ETH_HandleTypeDef *hmdio = &dev_data->hmdio;
+	uint32_t read;
+	int ret;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	ret = HAL_ETH_ReadPHYRegister(hmdio, prtad, regad, &read);
+
+	k_sem_give(&dev_data->sem);
+
+	if (ret != HAL_OK)
+		return -EIO;
+
+	*data = read & 0xffff;
```

Comment:
use GENMASK macro in zephyr/sys/util.h preferably or FIELD_GET

Reply:
thanks, done